src/types.rs

```diff
@@ -450,6 +459,52 @@
     }
 }
 
+pub struct SubjectBuilder {
+    tokens: Vec<String>
+}
+
+impl SubjectBuilder {
+    pub fn new() -> Self {
+        SubjectBuilder {
+            tokens: Vec::new()
+        }
+    }
+
+    pub fn add(mut self, subject: String) -> Self {
+        // Need to add some checks here to check for illegal characters
+        self.tokens.push(subject);
+        self
+    }
+
+    pub fn add_wildcard(mut self) -> Self {
+        self.tokens.push("*".to_string());
+        self
+    }
+
+    pub fn add_full_wildcard(self) -> FullWildcardSubjectBuilder {
+        FullWildcardSubjectBuilder (self)
+    }
+
+    pub fn build(self) -> Subject {
+        let fwc = self.tokens.is_empty();
```

Comment:
Nice! I like that we guarentee that the `Subject` is always valid by making it a full wildcard if no tokens were specified.